kubespawner/proxy.py

```diff
@@ -0,0 +1,244 @@
+                    body=body
+                )
+                self.log.info('Created %s/%s', kind, safe_name)
+            except client.rest.ApiException as e:
+                if e.status == 409:
+                    # This object already exists, we should patch it to make it be what we want
+                    self.log.warn("Trying to patch %s/%s, it already exists", kind, safe_name)
+                    resp = yield self.asynchronize(
+                        patch_func,
+                        namespace=self.namespace,
+                        body=body,
+                        name=body.metadata.name
+                    )
+                else:
+                    raise
+
+        yield ensure_object(
+            self.core_api.create_namespaced_endpoints,
+            self.core_api.patch_namespaced_endpoints,
+            body=endpoint,
+            kind='endpoints'
+        )
+
+        yield exponential_backoff(
+            lambda: safe_name in self.endpoint_reflector.endpoints,
+            'Could not find endpoints/%s after creating it' % safe_name
+        )
+
+        yield ensure_object(
+            self.core_api.create_namespaced_service,
+            self.core_api.patch_namespaced_service,
+            body=service,
+            kind='service'
+        )
+
+        yield exponential_backoff(
+            lambda: safe_name in self.service_reflector.services,
+            'Could not find service/%s after creating it' % safe_name
+        )
+
+        yield ensure_object(
+            self.extension_api.create_namespaced_ingress,
+            self.extension_api.patch_namespaced_ingress,
+            body=ingress,
+            kind='ingress'
+        )
+
+        yield exponential_backoff(
+            lambda: safe_name in self.ingress_reflector.ingresses,
+            'Could not find ingress/%s after creating it' % safe_name
+        )
+
+    @gen.coroutine
+    def delete_route(self, routespec):
+        # We just ensure that these objects are deleted.
+        # This means if some of them are already deleted, we just let it
+        # be.
+        safe_name = self.safe_name_for_routespec(routespec).lower()
+
+        delete_options = client.V1DeleteOptions(grace_period_seconds=0)
+
+        delete_endpoint = self.asynchronize(
+            self.core_api.delete_namespaced_endpoints,
+            name=safe_name,
+            namespace=self.namespace,
+        )
+
+        delete_service = self.asynchronize(
+            self.core_api.delete_namespaced_service,
+            name=safe_name,
+            namespace=self.namespace,
+        )
+
+        delete_ingress = self.asynchronize(
+                self.extension_api.delete_namespaced_ingress,
+                name=safe_name,
+                namespace=self.namespace,
+                body=delete_options,
+                grace_period_seconds=0
+        )
+
+        # This seems like cleanest way to parallelize all three of these while
```

Comment:
Question for my own understanding: does order of deletion matter at all (i.e is there any dependence on each other)?

Reply:
yup! Added a line here!

Reply:
It does! I've added a comment!